beluga_ros/include/beluga_ros/amcl_impl.hpp

```diff
@@ -0,0 +1,215 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ROS_AMCL_IMPL_HPP
+#define BELUGA_ROS_AMCL_IMPL_HPP
+
+#include <optional>
+#include <utility>
+#include <variant>
+
+#include <beluga/beluga.hpp>
+#include <beluga_ros/laser_scan.hpp>
+#include <beluga_ros/occupancy_grid.hpp>
+
+namespace beluga_ros {
+
+/// Struct containing parameters for the Adaptive Monte Carlo Localization (AMCL) implementation.
+struct AmclImplParams {
+  double update_min_d = 0.25;
+  double update_min_a = 0.2;
+  std::size_t resample_interval = 1UL;
+  bool selective_resampling = false;
+  std::size_t min_particles = 500UL;
+  std::size_t max_particles = 2000UL;
+  double alpha_slow = 0.001;
+  double alpha_fast = 0.1;
+  double kld_epsilon = 0.05;
+  double kld_z = 3.0;
+  double spatial_resolution_x = 0.5;
+  double spatial_resolution_y = 0.5;
+  double spatial_resolution_theta = 10 * Sophus::Constants<double>::pi() / 180;
+};
+
+/// Implementation of the Beluga Adaptive Monte Carlo Localization (AMCL) algorithm.
+class AmclImpl {
```

Comment:
@nahueespinosa meta: why `AmclImpl` instead of just `Amcl` or `AdaptiveMonteCarloLocalization`?

Reply:
I went with `Amcl`.